app/models/manageiq/providers/redhat/inventory/parser/infra_manager.rb

```diff
@@ -0,0 +1,515 @@
+      cluster = persister.clusters.lazy_find(cluster_id)
+      datacenter_id = cluster.dig(:data_center, :id)
+      network = networks.detect { |n| n.name == network_name && n.dig(:data_center, :id) == datacenter_id }
+    end
+
+    network
+  end
+
+  def lans(persister_switch, network)
+    tag_value = nil
+    if network
+      uid = network.id
+      name = network.name
+      tag_value = network.try(:vlan).try(:id)
+    else
+      uid = name = network_name unless network_name.nil?
+    end
+
+    if uid.nil?
+      return
+    end
+
+    persister.lans..find_or_build_by(
+      :switch  => persister_switch,
+      :uid_ems => uid
+    ).assign_attributes(
+      :name    => name,
+      :uid_ems => uid,
+      :tag     => tag_value
+    )
+  end
+
+  def host_operating_systems(persister_host, host, hostname)
+    persister.operating_systems.find_or_build_by(persister_host).assign_attributes(
+      :name         => hostname,
+      :product_type => 'linux',
+      :product_name => extract_host_os_name(host),
+      :version      => extract_host_os_full_version(host.os)
+    )
+  end
+
+  def vms
+    vms = Array(collector.vms) + Array(collector.templates)
+
+    vms.each do |vm|
+      # Skip the place holder template
+      next if vm.id == '00000000-0000-0000-0000-000000000000'
+
+      template = vm.href.include?('/templates/')
+
+      ems_ref = ManageIQ::Providers::Redhat::InfraManager.make_ems_ref(vm.href)
+
+      host_id = vm.try(:host).try(:id)
+      host_id = vm.try(:placement_policy).try(:hosts).try(:id) if host_id.blank?
+
+      storages = storages(vm)
+
+      persister_vm = persister.vms.find_or_build(vm.id).assign_attributes(
+        :type             => template ? "ManageIQ::Providers::Redhat::InfraManager::Template" : "ManageIQ::Providers::Redhat::InfraManager::Vm",
+        :ems_ref          => ems_ref,
+        :ems_ref_obj      => ems_ref,
+        :uid_ems          => vm.id,
+        :connection_state => "connected",
+        :vendor           => "redhat",
+        :name             => URI.decode(vm.name),
+        :location         => "#{vm.id}.ovf",
+        :template         => template,
+        :memory_reserve   => vm_memory_reserve(vm),
+        :raw_power_state  => template ? "never" : vm.status,
+        :boot_time        => vm.try(:start_time),
+        :host             => persister.hosts.lazy_find(host_id),
+        :ems_cluster      => persister.clusters.lazy_find(vm.cluster.id),
+        :storages         => storages,
+        :storage          => storages.first,
+      )
+
+      vm_hardware(persister_vm, vm)
+      snapshots(persister_vm, vm)
+      operating_systems(persister_vm, vm)
+      custom_attributes(persister_vm, vm)
+    end
+  end
+
+  def storages(vm)
+    storages = []
+    collector.attached_disks(vm).to_miq_a.each do |disk|
+      disk.storage_domains.to_miq_a.each do |sd|
+        storages << persister.storages.lazy_find(:ems_ref => sd.id)
```

Comment:
lazy find knows only strings now and can take only the unique uuid (if needed we could extend it to teach it multiple indexes)

so persister.storages.lazy_find(sd.id)